### PR TITLE
Prevent leaking goroutines from websocket connections

### PR DESCRIPTION
**Pull request by amerry**
The write to the CloseNotify() channel was blocking, preventing the
Read() call from exiting, and preventing ServeHTTP from returning. This
meant that requests tended to stay around forever in a deadlocked state.

Using the context package prevent this, as the cancellation is
non-blocking. It depends on Go 1.7.

Note that the writer still needs to support the CloseNotifier interface,
otherwise other things break, but it does not seem to be necessary to
signal using it.

### CODE REVIEW

File: go/grpcweb/wrapper.go [hunk 1]
```diff
@@ -148,14 +149,17 @@ func (w *WrappedGrpcServer) handleWebSocket(wsConn *websocket.Conn, req *http.Re
 		return
 	}
 
+	ctx, cancelFunc := context.WithCancel(req.Context())
+	defer cancelFunc()
```
Comment (johanbrandhorst) [inline go/grpcweb/wrapper.go]: Do we need to explicitly cancel this? Can we not rely on the built-in request context being cancelled?

Comment (johanbrandhorst) [reply go/grpcweb/wrapper.go]: If we can rely on the request context we don't need to pass around a cancelFunc either. I assume you've tested some of this?

Comment (amerry) [reply go/grpcweb/wrapper.go]: Yeah, we've got a websocket client that makes a request every second, so the memory leak is really obvious, and I put in some logging for testing (tracking the start and end of handleWebSocket, as well as before and after the cancellation step in the reader).

I tested a bunch of different configurations. Not cancelling the request context also prevented connections from closing properly; I'm not sure why (you would expect `w.server.ServeHTTP` to pick up from the error returned from the `Body` that it would need to cancel the request, but this doesn't seem to happen).

I'll admit that I didn't go diving into the lower-level code to see what was going on - this was mostly by trial-and-error.

Comment (johanbrandhorst) [reply go/grpcweb/wrapper.go]: OK - I'm not worried that this will cause any real problems, since the request is over once this function has exited, just thought it was a bit overzealous, but if your testing implies the necessity, I'm OK with it!

**Comment by johanbrandhorst** [comment]
Would like a confirmation from @MarcusLongmuir that we're OK requiring Go 1.7+ for building this now.

**Comment by amerry** [comment]
Oh, wait, I think I misunderstood. That `defer` there isn't necessary to prevent the goroutine leaking. I just put it in based on the `context.WithCancel` documentation, which seemed to suggest it was good practice.

**Comment by johanbrandhorst** [comment]
Let me rephrase: can you categorically confirm that _just_ removing the channel doesn't work? I imagine the request context will be cancelled when the client finishes reading.

**Comment by johanbrandhorst** [comment]
Going to give the last error another go :grimacing:

**Comment by johanbrandhorst** [comment]
Just going to give @MarcusLongmuir and @easyCZ a couple of days to put anymore comments on this but LGTM!

**Comment by johanbrandhorst** [comment]
OK I think we've had enough time to think about this, going to merge, thanks for your contribution!
